### PR TITLE
Feature/pings

### PR DESCRIPTION
**Pull request by connor4312**
This adds ping funtionality to the socket client to proactively detect disconnects/timeouts. Pings are sent, by default, if we go 15 seconds without any data from the socket and the server is given five seconds to respond before the client considers the connection to be dead. As a minor thing this also adds the ability to override various timing options in an optional object passed to the websocket constructor.

On our frontend it looks like this:

### CODE REVIEW

**Comment by connor4312** [comment]
This should solve the remaining two issues on https://trello.com/c/aNFA1JSl/48-test-how-chat-handles-reconnection-during-various-failure-stats-in-prod @MattJustMatt

File: lib/ws/ws.js [hunk 1]
```diff
@@ -9,13 +9,86 @@ var factory = require('./factory');
 var authMethod = 'auth';
 
 /**
+ * race takes an array of promises and is resolved with the first promise
+ * that resolves, or rejects. After the first resolve or rejection, all
+ * future resolutions and rejections will be discarded.
+ * @param {Promise[]} promises
+ * @return {Promise}
+ */
+function race(promises) {
```
Comment (JamyDev) [inline lib/ws/ws.js]: Why is this added manually? Bluebird has a `.race` [method](http://bluebirdjs.com/docs/api/promise.race.html) that seems to behave similarly.

Comment (connor4312) [inline lib/ws/ws.js]: We don't use Bluebird on the frontend.

Comment (JamyDev) [inline lib/ws/ws.js]: Then it would seem that we should polyfill that on frontend instead of here. Promises should have a race method [by specification](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/race).

Comment (connor4312) [inline lib/ws/ws.js]: Currently the Beam client is written so that it is agnostic and works with any A+ promise implementation, of which there are many on the Js world. There are fewer who are strict ES6 promise implementations, and RSVP which we use on the frontend is not one. Changing it here would introduce a breaking change for all client consumers, which I'd rather avoid doing.

Comment (JamyDev) [inline lib/ws/ws.js]: Sounds like a good enough reason to me 😄

**Comment by JamyDev** [comment]
LGTM

**Comment by MattJustMatt** [comment]
As discussed, we should make the reconnect interval random to avoid potential cascading failures due to 100,000+ reconnects all at once.

File: lib/ws/ws.js [hunk 2]
```diff
@@ -9,13 +9,86 @@ var factory = require('./factory');
...
+}
+
+/**
+ * Assigns enumerable properties of the src onto the obj, returning src.
+ * Semi-polyfill for Object.assign.
+ * @param  {Object} obj
+ * @param  {Object} src
+ * @return {Object}
+ */
+function assign(src, obj) {
```
Comment (ProbablePrime) [inline lib/ws/ws.js]: lodash is a dependency of this module, could we not use https://lodash.com/docs#assign

File: lib/ws/ws.js (hunk 1 above)
Comment (SimonSchick) [inline lib/ws/ws.js]: ```We don't use Bluebird on the frontend.```

Maybe we should change that.

File: lib/ws/ws.js (hunk 2 above)
Comment (connor4312) [inline lib/ws/ws.js]: Ah, yea, I forgot we had that 😛

File: lib/ws/ws.js (hunk 1 above)
Comment (connor4312) [inline lib/ws/ws.js]: Why's that?

File: test/unit/ws.test.js [hunk 3]
```diff
@@ -317,4 +317,84 @@ describe('websocket', function () {
...
+                    expect(err).to.be.an.instanceof(BeamSocket.TimeoutError);
+                    done();
+                });
+
+                clock.tick(5001);
+            });
+
+            it('should succeed if pong is received', function () {
+                socket.send.returns(Bluebird.resolve());
+                socket.on('error', function (err) {
```
Comment (SimonSchick) [inline test/unit/ws.test.js]: Shouldn't this be `socket.once('error', done)`?

File: lib/ws/ws.js (hunk 1 above)
Comment (connor4312) [inline lib/ws/ws.js]: fyi this was the reasoning originally against Bluebird:

> We use Bluebird on the backend, but 1) it's pretty heavy filesize-wise, 2) we don't really need a super micro-optimized promises lib on the frontend (that's Bluebird's claim to fame), and 3) those optimizations were made specifically for V8 and won't carry over for other browers.

Comment (SimonSchick) [inline lib/ws/ws.js]: Generally not a fan of using different libraries that do the same thing across multiple projects, it would also have prevented you having to re-inventing those functions.

However, if you really want to keep it agnostic, you may want to allow the end user of the library to inject their own promise implementation maybe via `require('beam-client-node').use('Promise', PROMISE_CONSTRUCTOR);`, this way you can drop the bluebird dependency.

Comment (connor4312) [inline lib/ws/ws.js]: We do allow the user to inject their own promise via setting BeamSocket.Promise -- that's what's done on the frontend. Using Bluebird would not have removed the need to reinvent this, since the API in the client only requires an A+ promise implementation.

Comment (SimonSchick) [inline lib/ws/ws.js]: Also `1) it's pretty heavy filesize-wise` you do realise bluebird is included in the frontend as it's a required dependency of beam-client-node.

I feel like we should remove bluebird as a dependency otherwise to gain full advantage of this abstraction.

Comment (connor4312) [inline lib/ws/ws.js]: Bluebird is a subdependency of several modules, including beam-client-node, but is excluded from the webpack bundle.

Comment (SimonSchick) [inline lib/ws/ws.js]: Also `Bluebird.race` works with any promise that implements `.then`, it's not specifically made for bluebird promises.

Comment (connor4312) [inline lib/ws/ws.js]: That's right, but Bluebird is not included in the webpack bundle.

**Comment by SimonSchick** [comment]
lgtm
